src/wix/WixToolset.Core.Burn/BurnCommand.cs

```diff
@@ -0,0 +1,78 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved. Licensed under the Microsoft Reciprocal License. See LICENSE.TXT file in the project root for full license information.
+
+namespace WixToolset.Core.Burn
+{
+    using System;
+    using System.Threading;
+    using System.Threading.Tasks;
+    using WixToolset.Extensibility.Data;
+    using WixToolset.Extensibility.Services;
+
+    /// <summary>
+    /// Burn specialized command.
+    /// </summary>
+    internal class BurnCommand : ICommandLineCommand
+    {
+        public BurnCommand(IServiceProvider serviceProvider)
+        {
+            this.ServiceProvider = serviceProvider;
+        }
+
+        public bool ShowHelp { get; set; }
+
+        public bool ShowLogo { get; set; }
+
+        public bool StopParsing { get; set; }
+
+        private IServiceProvider ServiceProvider { get; }
+
+        private BurnSubcommandBase Subcommand { get; set; }
+
+        public Task<int> ExecuteAsync(CancellationToken cancellationToken)
+        {
+            if (this.ShowHelp || this.Subcommand is null)
+            {
+                DisplayHelp();
+                return Task.FromResult(1);
+            }
+
+            return this.Subcommand.ExecuteAsync(cancellationToken);
+        }
+
+        public bool TryParseArgument(ICommandLineParser parser, string argument)
+        {
+            if (this.Subcommand is null)
+            {
+                switch (argument.ToLowerInvariant())
+                {
+                    case "detach":
+                        this.Subcommand = new DetachSubcommand(this.ServiceProvider);
+                        return true;
+
+                    case "reattach":
+                        this.Subcommand = new ReattachSubcommand(this.ServiceProvider);
+                        return true;
+                }
+
+                return false;
+            }
+
+            return this.Subcommand.TryParseArgument(parser, argument);
+        }
+
+        private static void DisplayHelp()
+        {
+            Console.WriteLine();
+            Console.WriteLine("Usage: wix burn detach|reattach bundle.exe -out engine.exe");
```

Comment:
Usage smells wrong, as `-out` isn't the engine, and it's missing some args, right?

Reply:
This isn't complete. I have an issue I'll open when this is ready to do a sweep through the whole help system. I've got a better feel for how I want things fit together now than when I originally wrote a lot of this cli code. There is a lot of "TBW" help strings.

Reply:
Fair. This comment was originally bigger -- before I saw that `-engine` was consistent across subcommands. (Originally I suggested adding `-bundle` to clarify but then decided `-out` made sense for the new bundle.)

Reply:
Yeah, I'm not married to these but after messing around a bit I think they made the most sense as I worked. Certainly made sense when I was writing the MSBuild task so that's a plus.